File: 03-sliding_window/00003-Longest_Substring_Without_Repeating_Characters/00003-longest_substring_without_repeating_characters.java
// 3: Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

import java.util.HashSet;

class Solution {
    // SOLUTION
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> letters = new HashSet<>();
        int i=0, j=0;
        int result = 0;

        for (; j<s.length(); j++) {
            while (letters.contains(s.charAt(j))) {
                letters.remove(s.charAt(i));
                i++;
            } 
            letters.add(s.charAt(j));
            result = Math.max(result, j-i+1);
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String s = "abcabcbb";

        // OUTPUT
        var result = o.lengthOfLongestSubstring(s);
        System.out.println(result);
    }
}

File: 03-sliding_window/00003-Longest_Substring_Without_Repeating_Characters/00003-longest_substring_without_repeating_characters.py
# 3: Longest Substring Without Repeating Characters
# https://leetcode.com/problems/longest-substring-without-repeating-characters/


class Solution:
    # SOLUTION
    def lengthOfLongestSubstring(self, s: str) -> int:
        letters = set()
        i, j = 0, 0
        result = 0

        for _ in range(len(s)):
            while s[j] in letters:
                letters.remove(s[i])
                i += 1

            letters.add(s[j])
            result = max(result, j-i+1)
            j += 1

        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    s: str = "abcabcbb"

    # OUTPUT
    result = o.lengthOfLongestSubstring(s)
    print(result)